File: NeetCode 150/Graphs/127. Word Ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        
        #Step 1: Build pattern_map
        pattern_map = defaultdict(list)
        L = len(beginWord)
        for word in wordList:
            for i in range(L):
                pattern = word[:i] + "*" + word[i+1:]
                pattern_map[pattern].append(word)
        
        #Step 2: BFS
        queue = deque([(beginWord,1)])
        visited = set([beginWord])

        while(queue):
            word,level = queue.popleft()
            for i in range(L):
                pattern = word[:i] + "*" + word[i+1:]

                for neighbor in pattern_map[pattern]:
                    if neighbor == endWord:
                        return level + 1

                    if neighbor not in visited:
                        visited.add(neighbor)
                        queue.append((neighbor,level+1))
                
                #Important Optimization : to avoid multiple checks and save time. For large inputs, we may get TLE
                pattern_map[pattern] = []
        return 0 

File: NeedCode 150 (JAVA)/Graphs/127. Word Ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordset  = new HashSet<>(wordList);
        if(!wordset.contains(endWord)) return 0;

        //Pattern -> List of words matching that pattern
        Map<String, List<String>> patternMap = new HashMap<>();

        //Preprocess all the words in the dictionary
        for(String word: wordList){
            addPatterns(word,patternMap);
        }

        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);

        Set<String> visited = new HashSet<>();
        visited.add(beginWord);

        int level = 1;

        while(!queue.isEmpty()){
            int size = queue.size();

            for(int s = 0; s<size; s++){
                String word = queue.poll();

                if(word.equals(endWord)) return level;

                char[] chars = word.toCharArray();

                for(int i = 0; i<chars.length; i++){
                    String pattern = word.substring(0,i) + "*" + word.substring(i+1);

                    List<String> neighbors = patternMap.getOrDefault(pattern,new ArrayList<>());

                    for(String neighbor: neighbors){
                        if(!visited.contains(neighbor)){
                            visited.add(neighbor);
                            queue.offer(neighbor);
                        }
                    }
                    //optional optimization :avoid re processing of this pattern
                    patternMap.put(pattern,new ArrayList<>());
                }
            }
            level++;
        }
        return 0;
    }

    private void addPatterns(String word, Map<String,List<String>> patternMap){
        for(int i = 0;i<word.length();i++){
            String pattern = word.substring(0,i) + "*" + word.substring(i+1);
            patternMap.computeIfAbsent(pattern, k-> new ArrayList<>()).add(word);
        }
    }
}
